### PR TITLE
Compilation with upstream

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7348/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe

File: src/uct/ib/base/ib_iface.c [hunk 1]
```diff
@@ -572,6 +576,34 @@ uct_ib_iface_res_domain_init(uct_ib_iface_res_domain_t *res_domain,
...
+    struct ibv_parent_domain_init_attr attr;
+    struct ibv_td_init_attr td_attr;
+
+    if (iface->super.worker->thread_mode == UCS_THREAD_MODE_MULTI) {
+        td_attr.comp_mask = 0;
+        res_domain->td = ibv_alloc_td(dev->ibv_context, &td_attr);
+        if (res_domain->td == NULL) {
+            ucs_error("ibv_alloc_td() on %s failed: %m",
+                      uct_ib_device_name(dev));
+            return UCS_ERR_IO_ERROR;
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c]: need to call `uct_ib_iface_res_domain_cleanup`, right?

File: src/uct/ib/base/ib_iface.c [hunk 2]
```diff
@@ -588,6 +620,20 @@ static void uct_ib_iface_res_domain_cleanup(uct_ib_iface_res_domain_t *res_domai
...
     }
+#elif HAVE_DECL_IBV_ALLOC_TD
+    int ret;
+
+    ret = ibv_dealloc_pd(res_domain->ibv_domain);
+    if (ret != 0) {
+        ucs_warn("ibv_dealloc_pd() failed: %m");
+        return;
+    }
+    if (res_domain->td != NULL) {
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c]: should not the TD be destroyed prior the parent domain?

File: src/uct/ib/mlx5/ib_mlx5_hw.h [hunk 3]
```diff
@@ -73,4 +75,6 @@ enum mlx5dv_obj_type {
     MLX5DV_OBJ_RWQ  = 1 << 3,
 };
 
+struct mlx5dv_context;
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_hw.h]: why is this needed?

File: src/uct/ib/mlx5/ib_mlx5_log.c [hunk 4]
```diff
@@ -218,6 +218,20 @@ static size_t uct_ib_mlx5_dump_dgram(char *buf, size_t max, void *seg, int is_et
...
+    if (qp_type == IBV_QPT_UD) {
+        return 1;
+    }
+#if HAVE_DECL_IBV_EXP_QPT_DC_INI
+    if (qp_type == IBV_EXP_QPT_DC_INI) {
+        return 1;
+    }
+#endif
+    return 0;
+
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_log.c]: extra line

File: src/uct/ib/ud/base/ud_iface.c [hunk 5]
```diff
@@ -253,8 +253,15 @@ uct_ud_iface_create_qp(uct_ud_iface_t *self, const uct_ud_iface_config_t *config
     /* TODO: exp attrs autoconf */
     struct ibv_exp_qp_init_attr qp_init_attr;
     struct ibv_qp_attr qp_attr;
+    struct ibv_pd *pd;
     int ret;
 
+    pd = uct_ib_iface_md(&self->super)->pd;
```
Comment (brminich) [inline src/uct/ib/ud/base/ud_iface.c]: maybe create a short inline function for that, since it is used twice (at least)?

File: test/gtest/uct/ib/test_ib_md.cc [hunk 6]
```diff
@@ -5,11 +5,10 @@
 * See file LICENSE for terms.
 */
 
-extern "C" {
 #include <uct/api/uct.h>
 #include <ucs/time/time.h>
 #include <uct/ib/base/ib_md.h>
-}
+
```
Comment (brminich) [inline test/gtest/uct/ib/test_ib_md.cc]: how is it related to the PR?

Comment (Artemy-Mellanox) [reply test/gtest/uct/ib/test_ib_md.cc]: This will not compile with upstream

**Comment by Artemy-Mellanox** [comment]
@brminich

File: src/uct/ib/base/ib_iface.c (hunk 1 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_iface.c]: Yes, this cleanup I've added: 
https://github.com/openucx/ucx/pull/2743/commits/053cd02e04a1fbb0ddd59cb688b4dd974e1e415f#diff-694c9dba586b3d5c9d63a14795983496R607

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7357/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7366/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_device.h [hunk 7]
```diff
@@ -113,12 +116,30 @@ typedef struct uct_ib_device_spec {
...
+    };
+} uct_ib_dv_attr_t;
+
 /**
  * IB device (corresponds to HCA)
  */
 typedef struct uct_ib_device {
     struct ibv_context          *ibv_context;    /* Verbs context */
     struct ibv_exp_device_attr  dev_attr;        /* Cached device attributes */
+    uct_ib_dv_attr_t            dv_attr;         /* Cached DV specific attributes */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h]: as mentioned before, need to remove this dv fields from ib common code

**Comment by yosefe** [comment]
also, we need to setup the jenkins slave with upstream driver and passing unit tests on it before we can merge this PR

File: config/m4/ib.m4 [hunk 8]
```diff
@@ -159,39 +169,46 @@ AS_IF([test "x$with_ib" == xyes],
...
+                               [[#include <infiniband/mlx5_hw.h>]])
+                       AC_CHECK_DECLS([
+                           IBV_EXP_QP_INIT_ATTR_RES_DOMAIN,
+                           IBV_EXP_RES_DOMAIN_THREAD_MODEL,
+                           ibv_exp_create_res_domain,
+                           ibv_exp_destroy_res_domain],
+                                [AC_DEFINE([HAVE_IBV_EXP_RES_DOMAIN], 1, [IB resource domain])
+                                 has_res_domain=yes], [], [[#include <infiniband/verbs_exp.h>]])
+                               ], [with_mlx5_hw=no
+                                   with_ib_hw_tm=no])
```
Comment (yosefe) [inline config/m4/ib.m4]: no need to disable with_ib_hw_tm, only with_mlx5_hw

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7409/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe (passed both on hpc-ucx-pr & ucx-test)

File: contrib/test_jenkins.sh [hunk 9]
```diff
@@ -46,7 +46,8 @@ fi
 #
 if [ -n "$EXECUTOR_NUMBER" ] && [ -n "$JENKINS_RUN_TESTS" ]
 then
-	AFFINITY="taskset -c $(( 2 * EXECUTOR_NUMBER ))","$(( 2 * EXECUTOR_NUMBER + 1))"
+        nproc=$(grep ^processor /proc/cpuinfo | wc -l)
```
Comment (yosefe) [inline contrib/test_jenkins.sh]: better enable HT on the server because core numbers are not always LSB

File: src/uct/ib/rc/base/rc_iface.c [hunk 10]
```diff
@@ -196,10 +196,13 @@ ucs_status_t uct_rc_iface_query(uct_rc_iface_t *iface,
         iface_attr->cap.flags              |= UCT_IFACE_FLAG_ATOMIC_DEVICE;
 
         iface_attr->cap.atomic64.op_flags  |= UCS_BIT(UCT_ATOMIC_OP_ADD);
-        iface_attr->cap.atomic64.fop_flags |= UCS_BIT(UCT_ATOMIC_OP_ADD)  |
-                                              UCS_BIT(UCT_ATOMIC_OP_CSWAP);
+        iface_attr->cap.atomic64.fop_flags |= UCS_BIT(UCT_ATOMIC_OP_ADD);
+#if HAVE_IB_EXT_ATOMICS
+        iface_attr->cap.atomic64.fop_flags |= UCS_BIT(UCT_ATOMIC_OP_CSWAP);
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: cswap 64bit is IBTA atomic. can be always supported

File: test/apps/test_ucx_tls.py [hunk 11]
```diff
@@ -21,36 +21,64 @@
...
-mlx5_am_roce = {
-       2 :      "rc_mlx5",
-      16 :      "rc_mlx5",
-      32 :      "rc_mlx5",
-      64 :      "rc_mlx5",
-     256 :      "ud_mlx5",
-    1024 :      "ud_mlx5",
- 1000000 :      "ud_mlx5",
-}
+if os.path.exists("/bin/ofed_info"):
```
Comment (yosefe) [inline test/apps/test_ucx_tls.py]: pls add comment that it's temp until ...

File: test/gtest/ucm/malloc_hook.cc [hunk 12]
```diff
@@ -30,49 +30,45 @@ extern "C" {
...
-            alloc_size += small_alloc_size;
-        }
+        status = ucm_set_event_handler(UCM_EVENT_VM_MAPPED,
+                0, mem_event_callback,
+                reinterpret_cast<void*>(this));
+        ASSERT_UCS_OK(status);
 
         for (;;) {
             void *ptr = malloc(small_alloc_size);
-            mi = mallinfo();
```
Comment (yosefe) [inline test/gtest/ucm/malloc_hook.cc]: which mallinfo is invoked with upstream? should be the one from ucm/ptmalloc286 if malloc hooks are installed

File: test/gtest/ucm/malloc_hook.cc [hunk 13]
```diff
@@ -30,49 +30,45 @@ extern "C" {
...
-                min_free_space = free_space;
                 break;
             } else {
                 m_pts.push_back(ptr);
-                alloc_size += small_alloc_size;
-                min_free_space = free_space;
             }
         }
+        ucm_unset_event_handler(UCM_EVENT_VM_MAPPED,
+                mem_event_callback,
```
Comment (yosefe) [inline test/gtest/ucm/malloc_hook.cc]: indent

File: test/gtest/ucm/malloc_hook.cc [hunk 14]
```diff
@@ -30,49 +30,45 @@ extern "C" {
...
-        alloc_size      = 0;
-        mi = mallinfo();
-        prev_free_space = mi.fsmblks + mi.fordblks;
-
-        while (alloc_size < (size_t)(prev_free_space * 0.90)) {
-            m_pts.push_back(malloc(small_alloc_size));
-            alloc_size += small_alloc_size;
-        }
+        status = ucm_set_event_handler(UCM_EVENT_VM_MAPPED,
+                0, mem_event_callback,
```
Comment (yosefe) [inline test/gtest/ucm/malloc_hook.cc]: indent

File: config/m4/ib.m4 [hunk 15]
```diff
@@ -292,13 +292,15 @@ AS_IF([test "x$with_ib" == xyes],
...
 
-       AS_IF([test "x$with_dc" != xno],
-           [AC_CHECK_DECLS(IBV_EXP_QPT_DC_INI, [], [with_dc=no], [[#include <infiniband/verbs.h>]])
-           AC_CHECK_MEMBERS([struct ibv_exp_dct_init_attr.inline_size], [] , [with_dc=no], [[#include <infiniband/verbs.h>]])
-           ])
-       AS_IF([test "x$with_dc" != xno],
-           [AC_DEFINE([HAVE_TL_DC], 1, [DC transport support])
-           transports="${transports},dc"])
+       AC_CHECK_DECLS([IBV_EXP_QPT_DC_INI],
+                [have_dc_exp=yes], [], [[#include <infiniband/verbs.h>]])
```
Comment (yosefe) [inline config/m4/ib.m4]: indent

File: config/m4/ib.m4 [hunk 16]
```diff
@@ -292,13 +292,15 @@ AS_IF([test "x$with_ib" == xyes],
...
+
+       AS_IF([test "x$with_dc" != xno -a "$have_dc_exp" ], [
```
Comment (yosefe) [inline config/m4/ib.m4]: x$have_dc_exp == xyes

File: config/m4/ib.m4 [hunk 17]
```diff
@@ -292,13 +292,15 @@ AS_IF([test "x$with_ib" == xyes],
...
+           AC_DEFINE([HAVE_TL_DC], 1, [DC transport support])
+           transports="${transports},dc"
+           AS_IF([test -n "$have_dc_exp"],
```
Comment (yosefe) [inline config/m4/ib.m4]: let's keep in this commit only the minimum required for upstream compilation, which for now would disable DC on upstream

File: src/uct/ib/base/ib_iface.h [hunk 18]
```diff
@@ -48,6 +48,13 @@ typedef enum {
     UCT_IB_DIR_NUM
 } uct_ib_dir_t;
 
+typedef enum {
+    UCT_IB_QPT_RC = IBV_QPT_RC,
+    UCT_IB_QPT_UD = IBV_QPT_UD,
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: need to define QPT for DC only , this enum can be nameless and pass around qp_type as int

File: config/m4/ib.m4 [hunk 19]
```diff
@@ -201,6 +201,8 @@ AS_IF([test "x$with_ib" == xyes],
                        AC_CHECK_DECLS([
                            mlx5dv_init_obj],
                                   [], [], [[#include <infiniband/mlx5dv.h>]])
+                       AC_CHECK_MEMBERS([struct mlx5dv_cq.cq_uar],
+                                  [], [], [[#include <infiniband/mlx5dv.h>]])
```
Comment (yosefe) [inline config/m4/ib.m4]: indent

File: src/uct/ib/mlx5/ib_mlx5_hw.h [hunk 20]
```diff
@@ -33,7 +33,9 @@ struct mlx5dv_cq {
     volatile uint32_t   *dbrec;
     uint32_t            cqe_cnt;
     uint32_t            cqe_size;
-    void                *uar;
+    void                *cq_uar;
+#undef HAVE_STRUCT_MLX5DV_CQ_CQ_UAR
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_hw.h]: pls add comment why it's done

File: config/m4/ib.m4 [hunk 21]
```diff
@@ -203,6 +203,8 @@ AS_IF([test "x$with_ib" == xyes],
                                   [], [], [[#include <infiniband/mlx5dv.h>]])
                        AC_CHECK_MEMBERS([struct mlx5dv_cq.cq_uar],
                                   [], [], [[#include <infiniband/mlx5dv.h>]])
+                       AC_CHECK_DECLS([ibv_alloc_td],
+                                  [has_res_domain=yes], [], [[#include <infiniband/verbs.h>]])
```
Comment (yosefe) [inline config/m4/ib.m4]: indent

File: src/uct/ib/base/ib_iface.h [hunk 22]
```diff
@@ -418,4 +424,18 @@ void uct_ib_iface_fill_ah_attr_from_addr(uct_ib_iface_t *iface,
     uct_ib_iface_fill_ah_attr_from_gid_lid(iface, lid, gid_p, path_bits, ah_attr);
 }
 
+static UCS_F_ALWAYS_INLINE
+struct ibv_pd *uct_ib_iface_pd(uct_ib_iface_t *iface)
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: uct_ib_iface_qp_pd

File: src/uct/ib/base/ib_iface.c [hunk 23]
```diff
@@ -432,6 +433,7 @@ static ucs_status_t uct_ib_iface_create_cq(uct_ib_iface_t *iface, int cq_length,
     int env_var_added = 0;
     int ret;
 
+    pthread_mutex_lock(&lock);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: let's try to avoid setenv() at all - set CQE_SIZE only if HAVE_DECL_IBV_EXP_SETENV
otherwise revert to 64b (dfl) CQE

File: src/uct/ib/rc/base/rc_iface.c (hunk 10 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/rc/base/rc_iface.c]: But it's implemented via EXP:
https://github.com/openucx/ucx/blob/master/src/uct/ib/rc/verbs/rc_verbs_ep.c#L422

Comment (yosefe) [reply src/uct/ib/rc/base/rc_iface.c]: it's SWAP, not CSWAP

File: src/uct/sm/self/self.c [hunk 24]
```diff
@@ -148,6 +148,10 @@ static UCS_CLASS_INIT_FUNC(uct_self_iface_t, uct_md_h md, uct_worker_h worker,
 {
     ucs_status_t status;
 
+    if (RUNNING_ON_VALGRIND) {
+        VALGRIND_MAKE_MEM_DEFINED(&self, sizeof(self));
```
Comment (hoopoepg) [inline src/uct/sm/self/self.c]: this macro may be used without RUNNING_ON_VALGRIND evaluation, afaik

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7447/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_iface.c [hunk 25]
```diff
@@ -492,6 +495,19 @@ static ucs_status_t uct_ib_iface_create_cq(uct_ib_iface_t *iface, int cq_length,
     }
 out:
     return status;
+#else
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: maybe the `#else` should be in line 475? to avoid code duplication of calling ibv_create_cq()

File: src/uct/sm/self/self.c (hunk 24 above)
Comment (yosefe) [reply src/uct/sm/self/self.c]: which field of "self" is undefined?

File: src/uct/ib/mlx5/ib_mlx5_log.c [hunk 26]
```diff
@@ -218,6 +218,19 @@ static size_t uct_ib_mlx5_dump_dgram(char *buf, size_t max, void *seg, int is_et
     }
 }
 
+static int is_qp_require_av_seg(int qp_type)
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_log.c]: uct_ib_mlx5_ prefix

File: src/uct/ib/base/ib_iface.c (hunk 25 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_iface.c]: I guess you mean #endif on line 475? It's problematic to pass coverity then - unsetenv becomes deadcode.

Comment (yosefe) [reply src/uct/ib/base/ib_iface.c]: we can put "env_var_added" under #ifdef.. would still be shorter

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7461/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7473/ for details (Mellanox internal link).
